### PR TITLE
doc: add sequence diagram for pre-linking flow

### PR DESCRIPTION
**Pull request by kleyow**

### CODE REVIEW

File: docs/sequence/PISPPreLinkingApi.puml [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  participant "inbound-server" as PISP_TP_IN
+end box
+box "Mojaloop"
+    participant Switch
+end box
+
+== Pre-Linking ==
+autonumber 1 "<b>PRELINKING-#</b>"
+
+PISP -> PISP_TP_OUT: GET /linking/get-dfsps
```
Comment (lewisdaly) [inline docs/sequence/PISPPreLinkingApi.puml:18]: I don't love this resource name... maybe we can just call it `GET /linking/dfsps` or `GET /linking/supportedDFSPS`

File: docs/sequence/PISPPreLinkingApi.puml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+rnote right of PISP_PLM: state: enabledDFSPLookupSuccess
+PISP_PLM -> PISP_TP_OUT: return TPEnabledDFSPs
+
+deactivate PISP_PLM
+
+PISP_TP_OUT --> PISP: 200 OK TPEnabledDFSPs
+
+note over PISP, Switch
+  ERROR_SCENARIO: Switch receives 7201 error.
+  NOTE: What microservice is handling service lookup?
```
Comment (lewisdaly) [inline docs/sequence/PISPPreLinkingApi.puml:69]: I had spoken to Michael about this - for now the 3p-api-adapter is fine, although that may change in the future as we expand the other dfsp roles

File: docs/sequence/PISPPreLinkingApi.puml [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    errorDescription: 'No thirdparty enabled FSP found'
+  }
+}
+end note
+rnote right of PISP_PLM: state: errored
+PISP_PLM -> PISP_TP_OUT: catch MojaloopFSPIOPError
+
+deactivate PISP_PLM
+
+PISP_TP_OUT --> PISP: 500 Internal Server Error ErrorInformationObject
```
Comment (lewisdaly) [inline docs/sequence/PISPPreLinkingApi.puml:108]: This should be a 404

**Review by lewisdaly — COMMENTED** [review]
Nice. I like that you separated out some of the errors. Just a few thoughts on resource names and sync http response codes.

File: docs/sequence/PISPPreLinkingApi.puml (hunk 1 above)
Comment (eoln) [reply docs/sequence/PISPPreLinkingApi.puml:18]: for discovery we are using `GET /linking/accounts/{dfspId}/{username}` so `GET /linking/dfsps` fits better for me

what about `GET /linking/providers` ?

File: docs/sequence/PISPPreLinkingApi.puml [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  HAPPY_SCENARIO: Switch returns a non-empty list
+end note
+
+activate PISP_TP_IN
+
+Switch --> PISP_TP_IN: PUT /services/THIRD_PARTY_DFSP
+rnote left of Switch #LightBlue
+PUT /services/THIRD_PARTY_DFSP
+{
+  serviceProviders: [
```
Comment (eoln) [inline docs/sequence/PISPPreLinkingApi.puml:46]:
```suggestion
   providers: [
```

File: docs/sequence/PISPPreLinkingApi.puml [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+deactivate PISP_TP_IN
+activate PISP_PLM
+
+rnote right of PISP_PLM: state: enabledDFSPLookupSuccess
+PISP_PLM -> PISP_TP_OUT: return TPEnabledDFSPs
+
+deactivate PISP_PLM
+
+PISP_TP_OUT --> PISP: 200 OK TPEnabledDFSPs
+
```
Comment (eoln) [inline docs/sequence/PISPPreLinkingApi.puml:66]: let show response body
```suggestion
rnote left of PISP #LightBlue
{  
    providers: [
       "dfspa", "dfspb"
    ],
    currentState="COMPLETED"
}
```

**Review by eoln — CHANGES_REQUESTED** [review]
I suggest also to split diagram in separate ones for error scenarios to inject some more text between

File: docs/sequence/PISPPreLinkingApi.puml (hunk 4 above)
Comment (kleyow) [reply docs/sequence/PISPPreLinkingApi.puml:46]: https://github.com/mojaloop/pisp/blob/master/docs/linking/README.md#142-otp has it as `serviceProviders`. Was this changed somewhere else?

makes sense though a bit redundant to have "service" repeated on the services resource.

File: docs/sequence/PISPPreLinkingApi.puml (hunk 1 above)
Comment (kleyow) [reply docs/sequence/PISPPreLinkingApi.puml]: `GET /linking/providers` sounds good.

File: docs/sequence/PISPPreLinkingApi.puml (hunk 2 above)
Comment (kleyow) [reply docs/sequence/PISPPreLinkingApi.puml]: thanks for the info. removed comment

File: docs/sequence/PISPPreLinkingApi.puml (hunk 3 above)
Comment (kleyow) [reply docs/sequence/PISPPreLinkingApi.puml]: thanks

File: docs/sequence/PISPPreLinkingApi.puml (hunk 5 above)
Comment (kleyow) [reply docs/sequence/PISPPreLinkingApi.puml]: done

**Comment by kleyow** [comment]
Split happy path and error scenarios into two as well.

File: docs/sequence/error_scenarios/PISPPreLinkingApi.puml [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+autonumber 1 "<b>PRELINKING-#</b>"
+
+PISP -> PISP_TP_OUT: GET /linking/providers
+
+activate PISP
+activate PISP_TP_OUT
+
+PISP_TP_OUT -> PISP_TP_OUT: const model = await create()
+rnote right of PISP_TP_OUT: state: start
+PISP_TP_OUT -> PISP_PLM: model.getThirdpartyEnabledDFSPs()
```
Comment (eoln) [inline docs/sequence/error_scenarios/PISPPreLinkingApi.puml:25]:
```suggestion
PISP_TP_OUT -> PISP_PLM: model.getProviders()
```

File: docs/sequence/PISPPreLinkingApi.puml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+autonumber 1 "<b>PRELINKING-#</b>"
+
+PISP -> PISP_TP_OUT: GET /linking/providers
+
+activate PISP
+activate PISP_TP_OUT
+
+PISP_TP_OUT -> PISP_TP_OUT: const model = await create()
+rnote right of PISP_TP_OUT: state: start
+PISP_TP_OUT -> PISP_PLM: model.getThirdpartyEnabledDFSPs()
```
Comment (eoln) [inline docs/sequence/PISPPreLinkingApi.puml:25]:
```suggestion
PISP_TP_OUT -> PISP_PLM: model.getProviders()
```

File: docs/sequence/PISPPreLinkingApi.puml [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+PISP_TP_IN --> Switch: 200 OK
+
+deactivate Switch
+
+PISP_TP_IN -> PISP_PLM: Services request recieved
+
+deactivate PISP_TP_IN
+activate PISP_PLM
+
+rnote right of PISP_PLM: state: enabledDFSPLookupSuccess
```
Comment (eoln) [inline docs/sequence/PISPPreLinkingApi.puml:60]:
```suggestion
rnote right of PISP_PLM: state: providersLookupSuccess
```

File: docs/sequence/PISPPreLinkingApi.puml [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+PISP_PLM -> PISP_TP_OUT: return TPEnabledDFSPs
```
Comment (eoln) [inline docs/sequence/PISPPreLinkingApi.puml:61]:
```suggestion
PISP_PLM -> PISP_TP_OUT: return providers
```

File: docs/sequence/PISPPreLinkingApi.puml [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+deactivate PISP_PLM
+rnote left of PISP_TP_OUT #LightBlue
+{
+    providers: [
+       "dfspa", "dfspb"
+    ],
+    currentState="enabledDFSPLookupSuccess"
```
Comment (eoln) [inline docs/sequence/PISPPreLinkingApi.puml:69]:
```suggestion
    currentState="providersLookupSuccess"
```

**Review by eoln — CHANGES_REQUESTED** [review]
I suggest to keep `providers` naming to be in sync with path

**Comment by kleyow** [comment]
addressed comments.

**Review by eoln — APPROVED** [review]
a great job is done @kleyow 
I am happy to see error scenarios materialising :-D

**Pull request merged**